### PR TITLE
[Flax] Add FlaxBlenderbot

### PR DESCRIPTION
**Pull request by stancld**
# What does this PR do?
This PR adds flax implementation of Blenderbot.

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

<hr>

### TODOs:

- fix PT-Flax model equivalence

## Who can review?

@patrickvonplaten  @patil-suraj

### CODE REVIEW

File: tests/test_modeling_flax_blenderbot.py [hunk 1]
```diff
@@ -0,0 +1,417 @@
...
+        cross_attn_head_mask = np.ones((config.decoder_layers, config.decoder_attention_heads))
+    return {
+        "input_ids": input_ids,
+        "decoder_input_ids": decoder_input_ids,
+        "attention_mask": attention_mask,
+        "decoder_attention_mask": attention_mask,
+    }
+
+
+class FlaxBartModelTester:
```
Comment (patrickvonplaten) [inline tests/test_modeling_flax_blenderbot.py:73]:
```suggestion
class FlaxBlenderbotModelTester:
```

**Comment by stancld** [comment]
@patrickvonplaten I would like to kindly ping for a review. :) I've been struggling to achieve the pt-flax equivalence, however, I cannot find that difference/bug in this new flax implementation.

Thanks a lot! :)

File: src/transformers/models/bart/modeling_flax_bart.py [hunk 2]
```diff
@@ -405,7 +405,7 @@ def setup(self) -> None:
             num_heads=self.config.encoder_attention_heads,
             dropout=self.config.attention_dropout,
         )
-        self.self_attn_layer_norm = nn.LayerNorm(dtype=self.dtype)
+        self.self_attn_layer_norm = nn.LayerNorm(dtype=self.dtype, epsilon=1e-05)
```
Comment (patrickvonplaten) [inline src/transformers/models/bart/modeling_flax_bart.py:408]: @patil-suraj - the default in PyTorch is 1e-05, so I adapted it for all Bart-like models. Given that PT and Flax tests were passing for Bart before I think this "bug correction" is fine in terms of backwards compatibility

**Review by patrickvonplaten — APPROVED** [review]
Good for me now!

**Comment by stancld** [comment]
@patrickvonplaten Thank you very much for spotting the problem! :]

**Comment by patrickvonplaten** [comment]
Tests on master seem to be broken currently :-/ 

But I think the PR is good to go. @patil-suraj could you maybe take a look once you're back (and maybe rebase to master with @stancld to fix the circli ci runner)

**Comment by patrickvonplaten** [comment]
Awesome - I let you merge @patil-suraj once you're back :-)

File: src/transformers/models/blenderbot/modeling_flax_blenderbot.py [hunk 3]
```diff
@@ -0,0 +1,1499 @@
...
+                f"embed_dim must be divisible by num_heads (got `embed_dim`: {self.embed_dim}"
+                f" and `num_heads`: {self.num_heads})."
+            )
+
+        dense = partial(
+            nn.Dense,
+            self.embed_dim,
+            use_bias=self.bias,
+            dtype=self.dtype,
+            kernel_init=jax.nn.initializers.normal(self.config.init_std, self.dtype),
```
Comment (patil-suraj) [inline src/transformers/models/blenderbot/modeling_flax_blenderbot.py:251]: We should not pass the `dtype` anymore to the `kernel_init`, it's meant to specify the `dtype` of computation and not of parameters. This was a bug in all flax models, which is fixed by #13098. 

@stancld Could you please rebase the branch again with master and fix this according to what is explained in #13098?

Comment (stancld) [reply src/transformers/models/blenderbot/modeling_flax_blenderbot.py]: @patil-suraj Thank you for providing me with the context. Should be fixed now :]

File: src/transformers/testing_utils.py [hunk 4]
```diff
@@ -463,6 +463,13 @@ def require_torch_tpu(test_case):
 if is_tf_available():
     import tensorflow as tf
 
+if is_flax_available():
+    import jax
+
+    jax_device = jax.default_backend()
+else:
+    jax_device = None
+
```
Comment (patil-suraj) [inline src/transformers/testing_utils.py:472]: nice!

**Review by patil-suraj — APPROVED** [review]
Thank you @stancld  for adding this, LGTM!

WIll push a couple of flax checkpoint and then merge :)

File: tests/test_modeling_flax_blenderbot.py [hunk 5]
```diff
@@ -0,0 +1,417 @@
...
+                        outputs = decode_jitted(**prepared_inputs_dict).to_tuple()
+
+                self.assertEqual(len(outputs), len(jitted_outputs))
+                for jitted_output, output in zip(jitted_outputs, outputs):
+                    self.assertEqual(jitted_output.shape, output.shape)
+
+    @slow
+    def test_model_from_pretrained(self):
+        for model_class_name in self.all_model_classes:
+            model = model_class_name.from_pretrained("facebook/blenderbot-400M-distill", from_pt=True)
```
Comment (patil-suraj) [inline tests/test_modeling_flax_blenderbot.py:395]:
```suggestion
            model = model_class_name.from_pretrained("facebook/blenderbot-400M-distill")
```

**Pull request merged**